tests/AbstractJsonProducerTest.php

```diff
@@ -399,4 +401,45 @@ public function it_uses_return_callback()
         $this->assertInstanceOf(Envelope::class, $result[1][4]);
         $this->assertEquals(['foo' => 'bar'], json_decode($result[1][5], true));
     }
+
+    /**
+     * @test
+     */
+    public function it_throws_exception_when_data_could_not_be_encoded_to_json()
+    {
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectExceptionMessage('Error during json encoding');
+
+        $exchange = $this->prophesize(Exchange::class);
+        $exchange->publish(Argument::any(), Argument::any(), Argument::any(), Argument::any())->shouldNotBeCalled();
+
+        $producer = new JsonProducer($exchange->reveal());
+
+        $text = "\xB1\x31";
+
+        $producer->publish($text);
+    }
+
+    /**
+     * @test
+     */
+    public function it_throws_exception_when_data_could_not_be_encoded_to_json_2()
+    {
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectExceptionMessage('Exception during json encoding');
+
+        $exchange = $this->prophesize(Exchange::class);
+        $exchange->publish(Argument::any(), Argument::any(), Argument::any(), Argument::any())->shouldNotBeCalled();
+
+        $producer = new JsonProducer($exchange->reveal());
+
+        $message = new class() implements \JsonSerializable {
+            public function jsonSerialize()
```

Comment:
As suggested by php PHP Doc's comments this will be catched and consumed by `json_encode`. Not saying the test is wrong, just _think_ they both get handled by the `json_last_error` condition.